File: langs/java/src/main/java/dev/dotlyte/Loader.java
package dev.dotlyte;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.*;
import java.util.stream.Collectors;

/**
 * Main loader orchestrator — v2.
 *
 * <p>Supports explicit file lists, upward directory walking, variable
 * interpolation, schema defaults/validation, overrides layer, and
 * system env-var blocklisting.</p>
 */
class Loader {

    // ── Source interface ────────────────────────────────────────

    /** A pluggable configuration source. */
    interface Source {
        String name();
        Map<String, Object> load(LoadOptions options);
    }

    // ── System env blocklist ───────────────────────────────────

    private static final Set<String> SYSTEM_ENV_BLOCKLIST;
    static {
        Set<String> s = new HashSet<>(Arrays.asList(
                "PATH", "HOME", "USER", "SHELL", "TERM", "LANG", "LOGNAME",
                "DISPLAY", "HOSTNAME", "EDITOR", "VISUAL", "PAGER", "SHLVL",
                "PWD", "OLDPWD", "TMPDIR", "COLORTERM", "TERM_PROGRAM",
                "_", "LS_COLORS", "XPC_SERVICE_NAME", "XPC_FLAGS",
                "MANPATH", "SSH_AUTH_SOCK", "ANDROID_HOME", "JAVA_HOME",
                "GOPATH", "GOROOT", "CARGO_HOME", "RUSTUP_HOME", "NVM_DIR",
                "PYENV_ROOT", "RBENV_ROOT", "GEM_HOME", "GEM_PATH"));
        SYSTEM_ENV_BLOCKLIST = Collections.unmodifiableSet(s);
    }

    private static final String[] SYSTEM_PREFIXES = {
            "npm_", "VSCODE_", "TERM_", "XDG_", "LC_", "LESS", "SSH_",
            "GPG_", "DBUS_", "DESKTOP_", "GDM_", "GNOME_", "GTK_",
            "WINDOWMANAGER", "SWAYSOCK"
    };

    private static boolean isSystemVar(String key) {
        if (SYSTEM_ENV_BLOCKLIST.contains(key)) return true;
        String upper = key.toUpperCase();
        for (String pfx : SYSTEM_PREFIXES) {
            if (upper.startsWith(pfx.toUpperCase())) return true;
        }
        return false;
    }

    // ── Root markers (findUp) ──────────────────────────────────

    private static final List<String> DEFAULT_ROOT_MARKERS = Arrays.asList(
            ".git", "package.json", "go.mod", "Cargo.toml", "pom.xml",
            "build.gradle", "build.gradle.kts", "pyproject.toml",
            "Gemfile", "composer.json", ".dotlyte"
    );

    // ── Fields ─────────────────────────────────────────────────

    private final LoadOptions options;

    Loader(LoadOptions options) {
        this.options = options;
    }

    // ── Main load ──────────────────────────────────────────────

    Config load() {
        Path baseDir = resolveBaseDir();
        List<Map<String, Object>> layers = new ArrayList<>();

        // 1 — Defaults (lowest priority)
        addIfPresent(layers, options.getDefaults());

        // 2 — Explicit files or auto-discovery
        if (options.getFiles() != null && !options.getFiles().isEmpty()) {
            for (String file : options.getFiles()) {
                Path path = baseDir.resolve(file);
                if (!Files.exists(path)) {
                    throw new DotlyteException.FileException(
                            path.toString(), "File not found");
                }
                addIfPresent(layers, parseFileByExtension(path));
            }
        } else if (options.getSources() != null) {
            for (String source : options.getSources()) {
                addIfPresent(layers, loadSource(source, baseDir));
            }
        } else {
            addIfPresent(layers, loadYamlFiles(baseDir));
            addIfPresent(layers, loadJsonFiles(baseDir));
            addIfPresent(layers, loadDotenvFiles(baseDir));
            addIfPresent(layers, loadEnvVars());
        }

        // 3 — Overrides (highest priority)
        addIfPresent(layers, options.getOverrides());

        // Merge
        Map<String, Object> merged = new LinkedHashMap<>();
        for (Map<String, Object> layer : layers) {
            merged = Merger.deepMerge(merged, layer);
        }

        // 4 — schema defaults
        Map<String, SchemaRule> schema = options.getSchema();
        if (schema != null) {
            Validator.applyDefaults(merged, schema);
        }

        // 5 — interpolation
        if (options.isInterpolateVars()) {
            merged = Interpolation.interpolateDeep(merged, merged);
        }

        // 6 — schema validation
        if (schema != null) {
            List<SchemaViolation> violations = Validator.validate(
                    merged, schema, options.isStrict());
            if (!violations.isEmpty()) {
                throw new DotlyteException.ValidationException(violations);
            }
        }

        // Build sensitive keys set
        Set<String> sensitiveKeys = new LinkedHashSet<>();
        if (schema != null) {
            sensitiveKeys.addAll(Validator.getSensitiveKeys(schema));
        }
        sensitiveKeys.addAll(Masking.buildSensitiveSet(merged, sensitiveKeys));

        return new Config(merged, schema, sensitiveKeys);
    }

    // ── Directory resolution ───────────────────────────────────

    private Path resolveBaseDir() {
        Path start;
        if (options.getCwd() != null) {
            start = Path.of(options.getCwd());
        } else {
            start = Path.of(System.getProperty("user.dir"));
        }

        if (!options.isFindUp()) return start;

        List<String> markers = options.getRootMarkers() != null
                ? options.getRootMarkers() : DEFAULT_ROOT_MARKERS;

        Path dir = start;
        while (dir != null) {
            for (String marker : markers) {
                if (Files.exists(dir.resolve(marker))) {
                    return dir;
                }
            }
            dir = dir.getParent();
        }
        return start;
    }

    // ── Source routing ──────────────────────────────────────────

    private Map<String, Object> loadSource(String name, Path baseDir) {
        switch (name) {
            case "defaults": return options.getDefaults();
            case "yaml":     return loadYamlFiles(baseDir);
            case "json":     return loadJsonFiles(baseDir);
            case "dotenv":   return loadDotenvFiles(baseDir);
            case "env":      return loadEnvVars();
            default:         return Collections.emptyMap();
        }
    }

    // ── Dotenv files ───────────────────────────────────────────

    private Map<String, Object> loadDotenvFiles(Path baseDir) {
        List<String> candidates = new ArrayList<>();
        candidates.add(".env");
        if (options.getEnv() != null) {
            candidates.add(".env." + options.getEnv());
        }
        candidates.add(".env.local");

        Map<String, Object> merged = new LinkedHashMap<>();
        for (String filename : candidates) {
            Path path = baseDir.resolve(filename);
            if (Files.exists(path)) {
                merged = Merger.deepMerge(merged, parseDotenvFile(path));
            }
        }
        return merged;
    }

    // ── Enhanced dotenv parser ──────────────────────────────────

    /**
     * Parse a .env file — supports multiline double-quoted values,
     * inline comments, escape sequences, and `export` prefix.
     */
    static Map<String, Object> parseDotenvFile(Path path) {
        Map<String, Object> result = new LinkedHashMap<>();
        try {
            String content = Files.readString(path);
            parseDotenvContent(content, result, path.toString());
        } catch (IOException e) {
            // Ignore unreadable .env files
        }
        return result;
    }

    /**
     * Parse raw dotenv content — used for tests and encryption decryption.
     */
    static Map<String, Object> parseDotenvContent(String content) {
        Map<String, Object> result = new LinkedHashMap<>();
        parseDotenvContent(content, result, "<raw>");
        return result;
    }

    @SuppressWarnings("unchecked")
    private static void parseDotenvContent(String content,
                                           Map<String, Object> result,
                                           String source) {
        String[] lines = content.split("\n", -1);
        int i = 0;
        while (i < lines.length) {
            String line = lines[i].trim();
            i++;

            if (line.isEmpty() || line.startsWith("#")) continue;
            if (line.startsWith("export ")) line = line.substring(7).trim();

            int eqIdx = line.indexOf('=');
            if (eqIdx == -1) continue;

            String key = line.substring(0, eqIdx).trim();
            String rawVal = line.substring(eqIdx + 1).trim();

            String value;
            if (rawVal.startsWith("\"")) {
                // Possibly multiline double-quoted
                StringBuilder sb = new StringBuilder();
                String rest = rawVal.substring(1);
                while (true) {
                    int closeIdx = findUnescapedQuote(rest);
                    if (closeIdx >= 0) {
                        sb.append(rest, 0, closeIdx);
                        break;
                    }
                    sb.append(rest).append("\n");
                    if (i >= lines.length) break;
                    rest = lines[i];
                    i++;
                }
                value = processEscapes(sb.toString());
            } else if (rawVal.startsWith("'")) {
                // Single-quoted: literal
                if (rawVal.length() >= 2 && rawVal.endsWith("'")) {
                    value = rawVal.substring(1, rawVal.length() - 1);
                } else {
                    value = rawVal.substring(1);
                }
            } else {
                // Unquoted — strip inline comment
                int hashIdx = rawVal.indexOf(" #");
                if (hashIdx >= 0) {
                    rawVal = rawVal.substring(0, hashIdx);
                }
                value = rawVal.trim();
            }

            result.put(key.toLowerCase(), Coercion.coerce(value));
        }
    }

    private static int findUnescapedQuote(String s) {
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c == '\\') { i++; continue; }
            if (c == '"') return i;
        }
        return -1;
    }

    private static String processEscapes(String s) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '\\' && i + 1 < s.length()) {
                char next = s.charAt(i + 1);
                switch (next) {
                    case 'n':  sb.append('\n'); i++; break;
                    case 'r':  sb.append('\r'); i++; break;
                    case 't':  sb.append('\t'); i++; break;
                    case '"':  sb.append('"');  i++; break;
                    case '\\': sb.append('\\'); i++; break;
                    default:   sb.append('\\');      break;
                }
            } else {
                sb.append(s.charAt(i));
            }
        }
        return sb.toString();
    }

    // ── YAML files ─────────────────────────────────────────────

    @SuppressWarnings("unchecked")
    private Map<String, Object> loadYamlFiles(Path baseDir) {
        try {
            List<String> candidates = new ArrayList<>(Arrays.asList("config.yaml", "config.yml"));
            if (options.getEnv() != null) {
                candidates.add("config." + options.getEnv() + ".yaml");
                candidates.add("config." + options.getEnv() + ".yml");
            }

            Map<String, Object> merged = new LinkedHashMap<>();
            org.yaml.snakeyaml.Yaml yaml = new org.yaml.snakeyaml.Yaml();

            for (String filename : candidates) {
                Path path = baseDir.resolve(filename);
                if (Files.exists(path)) {
                    String content = Files.readString(path);
                    Object data = yaml.load(content);
                    if (data instanceof Map) {
                        merged = Merger.deepMerge(merged, (Map<String, Object>) data);
                    }
                }
            }
            return merged;
        } catch (IOException e) {
            return Collections.emptyMap();
        }
    }

    // ── JSON files ─────────────────────────────────────────────

    @SuppressWarnings("unchecked")
    private Map<String, Object> loadJsonFiles(Path baseDir) {
        try {
            List<String> candidates = new ArrayList<>();
            candidates.add("config.json");
            if (options.getEnv() != null) {
                candidates.add("config." + options.getEnv() + ".json");
            }

            com.google.gson.Gson gson = new com.google.gson.Gson();
            Map<String, Object> merged = new LinkedHashMap<>();

            for (String filename : candidates) {
                Path path = baseDir.resolve(filename);
                if (Files.exists(path)) {
                    String content = Files.readString(path);
                    Map<String, Object> data = gson.fromJson(content, LinkedHashMap.class);
                    if (data != null) {
                        merged = Merger.deepMerge(merged, data);
                    }
                }
            }
            return merged;
        } catch (IOException e) {
            return Collections.emptyMap();
        }
    }

    // ── Environment variables ──────────────────────────────────

    private Map<String, Object> loadEnvVars() {
        Map<String, Object> result = new LinkedHashMap<>();
        String prefix = options.getPrefix();
        String pfx = prefix != null ? prefix.toUpperCase() + "_" : null;

        for (Map.Entry<String, String> entry : System.getenv().entrySet()) {
            String key = entry.getKey();
            String value = entry.getValue();

            if (pfx != null) {
                if (!key.startsWith(pfx)) continue;
                String cleanKey = key.substring(pfx.length()).toLowerCase();
                setNested(result, cleanKey, Coercion.coerce(value));
            } else if (options.isAllowAllEnvVars()) {
                result.put(key.toLowerCase(), Coercion.coerce(value));
            } else {
                // Filter system vars
                if (!isSystemVar(key)) {
                    result.put(key.toLowerCase(), Coercion.coerce(value));
                }
            }
        }
        return result;
    }

    // ── Generic file parser ────────────────────────────────────

    @SuppressWarnings("unchecked")
    private Map<String, Object> parseFileByExtension(Path path) {
        String name = path.getFileName().toString().toLowerCase();
        try {
            if (name.endsWith(".yaml") || name.endsWith(".yml")) {
                String content = Files.readString(path);
                Object data = new org.yaml.snakeyaml.Yaml().load(content);
                return data instanceof Map ? (Map<String, Object>) data : Collections.emptyMap();
            } else if (name.endsWith(".json")) {
                String content = Files.readString(path);
                Map<String, Object> data = new com.google.gson.Gson()
                        .fromJson(content, LinkedHashMap.class);
                return data != null ? data : Collections.emptyMap();
            } else if (name.endsWith(".env") || name.startsWith(".env")) {
                return parseDotenvFile(path);
            } else {
                // Try dotenv by default
                return parseDotenvFile(path);
            }
        } catch (IOException e) {
            throw new DotlyteException.FileException(
                    path.toString(), e.getMessage(), e);
        }
    }

    // ── Helpers ─────────────────────────────────────────────────

    @SuppressWarnings("unchecked")
    static void setNested(Map<String, Object> data, String key, Object value) {
        String[] parts = key.split("_");
        Map<String, Object> current = data;

        for (int i = 0; i < parts.length - 1; i++) {
            current.putIfAbsent(parts[i], new LinkedHashMap<>());
            Object next = current.get(parts[i]);
            if (next instanceof Map) {
                current = (Map<String, Object>) next;
            } else {
                Map<String, Object> newMap = new LinkedHashMap<>();
                current.put(parts[i], newMap);
                current = newMap;
            }
        }
        current.put(parts[parts.length - 1], value);
    }

    private void addIfPresent(List<Map<String, Object>> layers,
                              Map<String, Object> data) {
        if (data != null && !data.isEmpty()) {
            layers.add(data);
        }
    }
}

File: langs/php/src/Loader.php
<?php

declare(strict_types=1);

namespace Dotlyte;

/**
 * Main loader orchestrator (v2).
 */
final class Loader
{
    private const SYSTEM_ENV_BLOCKLIST = [
        'PATH', 'HOME', 'USER', 'SHELL', 'TERM', 'LANG', 'LC_ALL',
        'LOGNAME', 'HOSTNAME', 'PWD', 'OLDPWD', 'SHLVL', 'TMPDIR',
        'EDITOR', 'VISUAL', 'PAGER', 'DISPLAY',
        'SSH_AUTH_SOCK', 'SSH_AGENT_PID', 'GPG_AGENT_INFO',
        'COLORTERM', 'TERM_PROGRAM', 'TERM_PROGRAM_VERSION',
        'XPC_FLAGS', 'XPC_SERVICE_NAME', 'COMMAND_MODE',
        'LS_COLORS', 'LSCOLORS', 'CLICOLOR', 'GREP_OPTIONS',
        'COMP_WORDBREAKS', 'HISTSIZE', 'HISTFILESIZE', 'HISTCONTROL',
    ];

    private const SYSTEM_PREFIXES = [
        'npm_', 'VSCODE_', 'ELECTRON_', 'CHROME_', 'GITHUB_', 'CI_',
        'GITLAB_', 'JENKINS_', 'TRAVIS_', 'CIRCLECI_', 'HOMEBREW_',
        'JAVA_HOME', 'GOPATH', 'NVM_', 'RVM_', 'RBENV_', 'PYENV_',
        'CONDA_', 'VIRTUAL_ENV', 'CARGO_HOME',
    ];

    public function __construct(
        private readonly LoadOptions $options,
    ) {}

    public function load(): Config
    {
        $baseDir = $this->options->findUp ? $this->findBaseDir() : ($this->options->cwd ?? getcwd());
        $layers = [];

        if ($this->options->files !== null && !empty($this->options->files)) {
            // Explicit file mode
            foreach ($this->options->files as $f) {
                $full = $this->resolvePath($f, $baseDir);
                if (!file_exists($full)) {
                    throw new FileException(
                        "Config file not found: {$full}",
                        filePath: $full
                    );
                }
                $data = $this->parseFileByExtension($full);
                $this->appendIf($layers, $data);
            }
        } elseif ($this->options->sources !== null) {
            foreach ($this->options->sources as $source) {
                $data = $this->loadSource($source, $baseDir);
                $this->appendIf($layers, $data);
            }
        } else {
            $this->appendIf($layers, $this->options->defaults);
            $this->appendIf($layers, $this->loadYamlFiles($baseDir));
            $this->appendIf($layers, $this->loadJsonFiles($baseDir));
            $this->appendIf($layers, $this->loadDotenvFiles($baseDir));
            $this->appendIf($layers, $this->loadEnvVars());
        }

        // Overrides (highest priority)
        $this->appendIf($layers, $this->options->overrides);

        $merged = [];
        foreach ($layers as $layer) {
            $merged = Merger::deepMerge($merged, $layer);
        }

        // Interpolation
        if ($this->options->interpolateVars) {
            $merged = Interpolation::interpolateDeep($merged);
        }

        // Schema defaults
        if ($this->options->schema !== null) {
            Validator::applyDefaults($merged, $this->options->schema);
        }

        // Decryption
        $encKey = Encryption::resolveEncryptionKey($this->options->env);
        if ($encKey !== null) {
            $this->decryptRecursive($merged, $encKey);
        }

        // Schema validation
        if ($this->options->schema !== null && $this->options->strict) {
            Validator::assertValid($merged, $this->options->schema, $this->options->strict);
        }

        // Sensitive keys
        $sensitive = [];
        if ($this->options->schema !== null) {
            $sensitive = Validator::sensitiveKeys($this->options->schema);
        }
        $sensitive = array_merge($sensitive, Masking::buildSensitiveSet($merged));
        $sensitive = array_values(array_unique($sensitive));

        return new Config($merged, $this->options->schema, $sensitive);
    }

    /**
     * @param array<int, array<string, mixed>> $layers
     * @param array<string, mixed> $data
     */
    private function appendIf(array &$layers, array $data): void
    {
        if (!empty($data)) {
            $layers[] = $data;
        }
    }

    private function findBaseDir(): string
    {
        $dir = realpath($this->options->cwd ?? getcwd()) ?: getcwd();

        while (true) {
            foreach ($this->options->rootMarkers as $marker) {
                if (file_exists($dir . DIRECTORY_SEPARATOR . $marker)) {
                    return $dir;
                }
            }
            $parent = dirname($dir);
            if ($parent === $dir) {
                return $this->options->cwd ?? getcwd();
            }
            $dir = $parent;
        }
    }

    private function resolvePath(string $file, string $baseDir): string
    {
        if (str_starts_with($file, '/') || str_starts_with($file, DIRECTORY_SEPARATOR)) {
            return $file;
        }
        return $baseDir . DIRECTORY_SEPARATOR . $file;
    }

    /**
     * @return array<string, mixed>
     */
    private function loadSource(string $name, string $baseDir): array
    {
        return match ($name) {
            'defaults' => $this->options->defaults,
            'yaml' => $this->loadYamlFiles($baseDir),
            'json' => $this->loadJsonFiles($baseDir),
            'dotenv' => $this->loadDotenvFiles($baseDir),
            'env' => $this->loadEnvVars(),
            default => [],
        };
    }

    /**
     * @return array<string, mixed>
     */
    private function loadDotenvFiles(string $baseDir): array
    {
        $candidates = ['.env'];
        if ($this->options->env !== null) {
            $candidates[] = '.env.' . $this->options->env;
        }
        $candidates[] = '.env.local';

        $merged = [];
        foreach ($candidates as $filename) {
            $full = $baseDir . DIRECTORY_SEPARATOR . $filename;
            if (!file_exists($full)) {
                continue;
            }

            $data = $this->parseDotenv($full);
            $merged = Merger::deepMerge($merged, $data);
        }

        return $merged;
    }

    /**
     * @return array<string, mixed>
     */
    private function parseDotenv(string $filepath): array
    {
        $content = file_get_contents($filepath);
        if ($content === false) {
            return [];
        }

        $lines = explode("\n", $content);
        $result = [];
        $i = 0;
        $count = count($lines);

        while ($i < $count) {
            $line = trim($lines[$i]);
            $i++;

            if ($line === '' || str_starts_with($line, '#')) {
                continue;
            }

            // Strip optional export prefix
            if (str_starts_with($line, 'export ')) {
                $line = substr($line, 7);
            }

            $eqPos = strpos($line, '=');
            if ($eqPos === false) {
                continue;
            }

            $key = trim(substr($line, 0, $eqPos));
            $value = trim(substr($line, $eqPos + 1));

            // Quoted values
            if (strlen($value) >= 1 && in_array($value[0], ['"', "'", '`'], true)) {
                $quote = $value[0];
                if ($quote === "'" || $quote === '`') {
                    // Single-quoted: find closing
                    $endIdx = strpos($value, $quote, 1);
                    $value = $endIdx !== false ? substr($value, 1, $endIdx - 1) : substr($value, 1);
                } else {
                    // Double-quoted: may be multiline
                    $stripped = substr($value, 1);
                    $closingIdx = strpos($stripped, '"');
                    if ($closingIdx !== false) {
                        $value = $this->processEscapes(substr($stripped, 0, $closingIdx));
                    } else {
                        // Multiline
                        $buf = $stripped;
                        while ($i < $count) {
                            $buf .= "\n" . $lines[$i];
                            $i++;
                            $closingIdx = strrpos($buf, '"');
                            if ($closingIdx !== false) {
                                $value = $this->processEscapes(substr($buf, 0, $closingIdx));
                                break;
                            }
                        }
                    }
                }
            } else {
                // Unquoted — strip inline comment
                $commentIdx = strpos($value, ' #');
                if ($commentIdx !== false) {
                    $value = rtrim(substr($value, 0, $commentIdx));
                }
            }

            $result[strtolower($key)] = Coercion::coerce($value);
        }

        return $result;
    }

    private function processEscapes(string $s): string
    {
        return str_replace(
            ['\\n', '\\t', '\\r', '\\"', '\\\\'],
            ["\n", "\t", "\r", '"', '\\'],
            $s
        );
    }

    /**
     * @return array<string, mixed>
     */
    private function loadYamlFiles(string $baseDir): array
    {
        if (!function_exists('yaml_parse_file')) {
            return [];
        }

        $candidates = ['config.yaml', 'config.yml'];
        if ($this->options->env !== null) {
            $candidates[] = 'config.' . $this->options->env . '.yaml';
            $candidates[] = 'config.' . $this->options->env . '.yml';
        }

        $merged = [];
        foreach ($candidates as $filename) {
            $full = $baseDir . DIRECTORY_SEPARATOR . $filename;
            if (!file_exists($full)) {
                continue;
            }
            /** @var array<string, mixed>|false $data */
            $data = yaml_parse_file($full);
            if (is_array($data)) {
                $merged = Merger::deepMerge($merged, $data);
            }
        }

        return $merged;
    }

    /**
     * @return array<string, mixed>
     */
    private function loadJsonFiles(string $baseDir): array
    {
        $candidates = ['config.json'];
        if ($this->options->env !== null) {
            $candidates[] = 'config.' . $this->options->env . '.json';
        }

        $merged = [];
        foreach ($candidates as $filename) {
            $full = $baseDir . DIRECTORY_SEPARATOR . $filename;
            if (!file_exists($full)) {
                continue;
            }

            $content = file_get_contents($full);
            if ($content === false) {
                continue;
            }

            /** @var array<string, mixed>|null $data */
            $data = json_decode($content, true);
            if (is_array($data)) {
                $merged = Merger::deepMerge($merged, $data);
            }
        }

        return $merged;
    }

    /**
     * @return array<string, mixed>
     */
    private function loadEnvVars(): array
    {
        $result = [];
        $prefix = $this->options->prefix !== null
            ? strtoupper($this->options->prefix) . '_'
            : null;

        foreach (getenv() as $key => $value) {
            if (!is_string($value)) {
                continue;
            }

            if ($prefix !== null) {
                if (!str_starts_with($key, $prefix)) {
                    continue;
                }
                $cleanKey = strtolower(substr($key, strlen($prefix)));
                $coerced = Coercion::coerce($value);
                if ($coerced !== null) {
                    $this->setNested($result, $cleanKey, $coerced);
                }
            } elseif ($this->options->allowAllEnvVars) {
                $coerced = Coercion::coerce($value);
                if ($coerced !== null) {
                    $result[strtolower($key)] = $coerced;
                }
            } else {
                // Filter out system env vars
                if (in_array($key, self::SYSTEM_ENV_BLOCKLIST, true)) {
                    continue;
                }
                $skip = false;
                foreach (self::SYSTEM_PREFIXES as $pfx) {
                    if (str_starts_with($key, $pfx)) {
                        $skip = true;
                        break;
                    }
                }
                if (!$skip) {
                    $coerced = Coercion::coerce($value);
                    if ($coerced !== null) {
                        $result[strtolower($key)] = $coerced;
                    }
                }
            }
        }

        return $result;
    }

    /**
     * @param array<string, mixed> $data
     */
    private function setNested(array &$data, string $key, mixed $value): void
    {
        $parts = explode('_', $key);
        $current = &$data;

        for ($i = 0, $len = count($parts) - 1; $i < $len; $i++) {
            if (!isset($current[$parts[$i]]) || !is_array($current[$parts[$i]])) {
                $current[$parts[$i]] = [];
            }
            $current = &$current[$parts[$i]];
        }

        $current[$parts[count($parts) - 1]] = $value;
    }

    /**
     * @return array<string, mixed>
     */
    private function parseFileByExtension(string $fullPath): array
    {
        $ext = strtolower(pathinfo($fullPath, PATHINFO_EXTENSION));

        return match ($ext) {
            'env' => $this->parseDotenv($fullPath),
            'yaml', 'yml' => function_exists('yaml_parse_file')
                ? (is_array($d = yaml_parse_file($fullPath)) ? $d : [])
                : [],
            'json' => is_array($d = json_decode(file_get_contents($fullPath) ?: '', true)) ? $d : [],
            default => $this->parseDotenv($fullPath), // fallback to dotenv
        };
    }

    /**
     * @param array<string, mixed> $data
     */
    private function decryptRecursive(array &$data, string $keyHex): void
    {
        foreach ($data as $k => &$v) {
            if (is_array($v)) {
                $this->decryptRecursive($v, $keyHex);
            } elseif (Encryption::isEncrypted($v)) {
                $v = Coercion::coerce(Encryption::decryptValue($v, $keyHex));
            }
        }
    }
}
